inst/include/dust/dust.hpp

```diff
@@ -282,15 +283,45 @@ class Dust {
   // contents of the particle state (uses the set_state() and swap()
   // methods on particles).
   void reorder(const std::vector<size_t>& index) {
+#ifdef _OPENMP
+    #pragma omp parallel for schedule(static) num_threads(_n_threads)
+#endif
     for (size_t i = 0; i < _particles.size(); ++i) {
       size_t j = index[i];
       _particles[i].set_state(_particles[j]);
     }
-    for (auto& p : _particles) {
-      p.swap();
+#ifdef _OPENMP
+    #pragma omp parallel for schedule(static) num_threads(_n_threads)
+#endif
+    for (size_t i = 0; i < _particles.size(); ++i) {
+      _particles[i].swap();
     }
   }
 
+  std::vector<size_t> resample(const std::vector<real_t>& weights) {
+    std::vector<size_t> idx(n_particles());
+    auto it_weights = weights.begin();
+    auto it_idx = idx.begin();
+    if (_n_pars == 0) {
+      const size_t np = _particles.size();
+      real_t u = dust::unif_rand(_rng.state(0));
```

Comment:
Taking the first RNG state seems like it should be fine (and below) as we're unlikely to generate huge amounts of numbers here, and the number of times they're used between particles varies anyway.

I think there's also an argument for having a separate RNG for particle filter purposes: would avoid needing to copy this state to and from the device in the CUDA code (it would stay on the host only); keeps model simulations repeatable with more independence from how they have been run.
It would make the OPENMP block below less neat though.
In the CUDA, if we're doing it this way, I could also separate out the update_host/device functions to just do the RNG state.

What do you think?

Reply:
I think we can change to add auxillary rng state if needed later; it'd be a small change. One possibility is that we add one additional rng here for extra work and never copy that onto the device, then work out how to get that working in serial with the openmp bits